File: src/ModelLib/Face.java
package ModelLib;

import java.util.Vector;

import static ModelLib.Utils.cross;
import static ModelLib.Utils.normalize;

public class Face
{
    public int[] vids = new int[3];
    public Vector<Neighbor> neighbors;
    public Point normal;
    public Point c;
    public int label;

    public Face()
    {
        label = 0;
    }

    public Face(Vertex v1, Vertex v2, Vertex v3)
    {
        c = (v1.p.sum(v2.p.sum(v3.p))).div(3.0);
        normal = cross(v2.p.sub(v1.p), v3.p.sub(v1.p));
        normalize(normal);
        //if (dot(normal, c) < 0) normal = normal * (-1.0);
        label = 0;
    }

    public Face(Face f)
    {
        vids = f.vids;
        neighbors = f.neighbors;
        normal = f.normal;
        c = f.c;
        label = f.label;
    }
}

File: src/model/Face.cpp
//
// Created by timofey on 07.10.2021.
//

#include "../../include/model/Face.h"
#include "../../include/util/Util.h"

Face::Face(int v1, int v2, int v3)
{
    indices.x = v1;
    indices.y = v2;
    indices.z = v3;
}

bool operator==(const Face &a, const Face &b)
{
    if (a.indices.x == b.indices.x && a.indices.y == b.indices.y && a.indices.z == b.indices.z)
        return true;
    return false;
}

Face::Face(const Face &an)
{
    indices.x = an.indices.x;
    indices.y = an.indices.y;
    indices.z = an.indices.z;

    for (auto& f : an.neighbors)
        neighbors.push_back(f);

    normal = Point(an.normal);

    label = an.label;
}

Face::Face(Indices index, const Point& px, const Point& py, const Point& pz)
{
    indices = index;
    center = (px + py + pz) / 3;
    normal = util::cross(py - px, pz - px);
    util::normalizeV(normal);
}
